### ISSUE TITLE
Idea: per-row TTL feature for CQL based on Alternator's TTL implementation

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#13000]
CQL's TTL implementation has two peculiarities that some users don't like:

1. The expiration time is defined for each cell instead of on a CQL row. So if a row is partially _updated_ it will end up with different parts expiring at different times, which is strange and unexpected. It is also not possible to extend the expiration time of an already-existing row without writing it entirely again. 
2. The user doesn't get a deletion event in CDC when a row (or parts of it) eventually expire. Some users think these events are very important - e.g., allowing them to perform some additional cleanup when a row gets deleted.

We already have issues proposing 1 and 2 above separately (see #11539 and #8380 respectively), but in this issue I'm proposing to reuse code which we already have - Alternator's TTL implementation - to achieve both goals together.

Alternator's TTL implementation defines a per-row TTL mechanism: The user designates a special expiration-time column that determines the expiration time of an entire row, and a background thread scans the table, looks for expired rows, and deletes them. These deletions behave like "normal" deletions, so they appear in CDC.

Currently, the Alternator TTL implementation is in the alternator subdirectory (`alternator/ttl.cc`), looks for Alternator tables and columns, and is started in main.cc only when Alternator is used. But it was always planned to be, in the future, usable also for CQL, so it should be relatively easy to move it to be a generic service which works for both Alternator and CQL. We even have some comments suggesting where these changes need to be done.

Note that reusing the Alternator TTL implementation will mean that in addition to the above two advantages, we will get two new disadvantages:

1. The expiration-scanning thread takes up performance, significantly more than CQL's existing TTL feature which is part of compactions that happen anyway whether or not TTL is used.
2. The deletion of a row may happen with significant delay after its desired expiration time (whereas in CQL's existing TTL, a cell becomes invisible at the precise moment that its TTL dictates).

**Comment by avikivity** [comment]
We can fix the second issue by filtering away rows that are TTLed out.

**Comment by nyh** [comment]
> We can fix the second issue by filtering away rows that are TTLed out.

Right. To do that we need to SELECT the expiration-time column in every read (even reads which didn't ask for it) and check if the item was expired.

There is an extra complication of what happens if the user wrote to the row after the expiration time. We want the new write to take effect and be visible. If we assume that timestamps have real time, the filtering can leave cells whose timestamps are newer than the expiration time, and also when we finally go about about deleting the row - the row tombstone should be back-timestamped to the expiration time. I'm not sure CDC would like that.

Also, regarding CDC: The benefit of this mechanism is that expired rows will reach CDC. But, it can reach it some relatively-long time after the item expires. Adding extra filtering as you proposed will only make this discrepancy worse - now the row will be become unreadable exactly on the expiration time, but it may take another hour until this row will reach CDC. Users may find this odd.

**Comment by dorlaor** [comment]
How does a regular CQL cell TTL get discovered? Through compaction?
Doesn't it suffer a similar delay as the proposed row tll? If not, can we
extend
CQL itself, with a collaboration with C*?

A wild option is to create a compaction enhancement/policy for TTL rows.
I assume here that the TTL periods are short - days and not years.
In such cases, when we scan the tables, we can split the sstables based
on the near TTL rows (let's say by days). It's probably not worth it as a
scan will still be required.

**Comment by nyh** [comment]
The issue https://github.com/scylladb/scylladb/issues/11539 has a discussion on the question on how exactly the user will provide the TTL on each CQL row.

One option is to do something analogous to what the DynamoDB API does:
* The user designates, via an option of `CREATE TABLE` or `ALTER TABLE`, one regular column (we can allow `int` or `timestamp` types, or both) that will be the expiration-time column.
* Alternatively (idea by @elcallio), the special column can be designated by having a special type, e.g., `EXPIRATION`, where we validate that only a single column may ever have this type.
* This designated column can then be written, or read, through completely unchanged CQL.
* This approach makes it very easy and natural to insert a row complete with its row-level expiration time, to change the expiration time of an existing row - without changing anything else in the row or changing only part of it. It's even easy to read the expiration time of an item. 

A second option suggested by @elcallio is to write the row's expiration time as the expiration time of the already existing CQL row marker:
* This will save storage space (on disk an in memory) over the need to have another regular column, although the disk saving should be be very small (modern sstables don't repeat the column names).
* Since only `INSERT` writes a row marker, this means only `INSERT` will be able to set the row's TTL. This is a bit strange, but I think complete acceptable. An `UPDATE` will be able to change data in a row but not to set TTL.
* Currently, `INSERT .. WITH TTL` adds the expiration time to the row marker and to all the cells. If we don't want it to add the expiration to all cells (which will cause them to expire using the classic CQL TTL feature!) we will need to invent a new syntax for this - e.g., `INSERT ... WITH ROW TTL`
* If we want the user to be able to read the TTL of the row marker, we need to do https://github.com/scylladb/scylladb/issues/14019.

What I don't like about the second option is that currently, putting a TTL on the row marker already has an existing well-defined, albeit very strange and not useful, semantics: The row marker itself should expire at the given time, while the rest of the cells just stay alive. With this idea we'll need to modify the low-level read code to change this semantics - and NOT expire the row marker.

A third option is to add the concept of a row-level TTL to our low-level storage - both in-memory and on-disk - in a new place. This is arguably the "best" solution, but will be the most invasive change, and given the simplicity of the first option (using a named column), I wonder if it's justified.

**Comment by avikivity** [comment]
I think a free-standing column is best, since the user has all the tools to manipulate it. Designate it via ALTER TABLE, not the type.

Syntax can be

```cql
CREATE TABLE tab (
    id int PRIMARY KEY,
    t text,
    expiration timestamp TTL
);
```

```cql
ALTER TABLE tab
ALTER COLUMN expiration NOT TTL
```

and back

```cql
ALTER TABLE tab
ALTER COLUMN expiration TTL
```

**Comment by mykaul** [comment]
The 2nd option doesn't allow updating the TTL, which sounds to me like a common reasonable request?

**Comment by nyh** [comment]
Actually in the second option (@elcallio's suggestion to use the row-marker's TTL) you can change just the row marker's TTL by doing an `INSERT` operation without any data. This will supposedly insert an empty row, but what it really does is just write a new row marker and do nothing else (in particular, it doesn't clear the old data). So this will work, but I think it makes the user experience very awkward, and I personally prefer the first option of putting the new expiration time feature in a real named column.

**Comment by nyh** [comment]
> I think a free-standing column is best, since the user has all the tools to manipulate it. Designate it via ALTER TABLE, not the type.
> 
> Syntax can be
> 
> CREATE TABLE tab (
>     id int PRIMARY KEY,
>     t text,
>     expiration timestamp TTL
> );

I like this syntax, and already started to implement it.

For the type of the column I'm thinking of allowing either `timestamp` (recommended, but has excessive millisecond resolution) or `bigint` types. I could have also allowed other types like `int` but if I don't pretend it is an unsigned in (which Scylla doesn't have) we will face the 2038 bug, so I think it's a bad idea to allow.

**Comment by nyh** [comment]
> ALTER TABLE tab
> ALTER COLUMN expiration NOT TTL
> 
> and back
> 
> ALTER TABLE tab
> ALTER COLUMN expiration TTL

By the way, according to the documentation, the correct syntax for altering a column (currently, its type) doesn't have the keyword "COLUMN".  So these examples would look like

```cql
ALTER TABLE tab ALTER expiration NOT TTL
ALTER TABLE tab ALTER expiration TTL
```

Note that there will be no syntax to change the TTL column from col1 to col2 - you would need to first cancel col1's status as TTL, and then make col2 as the TTL. This is not a big loss, and in fact DynamoDB has exactly the same limitation - you can't use UpdateTimeToLive to enable TTL with col2 before you disable the old TTL with col1.

**Comment by nyh** [comment]
> 
> ALTER TABLE tab ALTER expiration NOT TTL
> ALTER TABLE tab ALTER expiration TTL
> 

@avikivity I have a slight dislike for this syntax, because of the double "ALTER" and also the unnecessary need to say which TTL column you're cancelling in the NOT TTL case. What do you think about this alternative syntax?

```
ALTER TABLE TTL expiration
ALTER TABLE NOT TTL
```
or for the cancelling one, maybe even more guessable as:
```
ALTER TABLE TTL NULL
```

**Comment by nyh** [comment]
Note to self:

Beyond the single-node functional tests that I'm already writing tons of (see work in progress in https://github.com/scylladb/scylladb/pull/28320), we'll need to write tests for this feature that check behavior in larger clusters. We have a small number of these tests for Alternator's TTL - such as a dtest that items expire even when a node is down, and a test/cluster test for checking the correct scheduling group is used for the expiration process (both scanning and deletions).

Although the Alternator tests will also provide the same guarantees for the CQL feature (since they both use the same underlying code), it will be better to have CQL versions of these tests. Moreover, we probably need more of these tests. One thing I think we missed testing is how the TTL expiration threads work on a multi-DC cluster.  The code in alternator/ttl.cc currently has a few FIXMEs regarding the multi-DC scenario, which we never went back to reconsider or test. We should.

**Comment by nyh** [comment]
> Note to self:
>...
> Although the Alternator tests will also provide the same guarantees for the CQL feature (since they both use the same underlying code), it will be better to have CQL versions of these tests.

Done. One of the patches https://github.com/scylladb/scylladb/pull/28320 is now test/cluster tests, two of them are based on an existing test/cluster and and dtest test for the Alternator TTL feature - but now testing the new CQL feature.

> Moreover, we probably need more of these tests. One thing I think we missed testing is how the TTL expiration threads work on a multi-DC cluster. The code in alternator/ttl.cc currently has a few FIXMEs regarding the multi-DC scenario, which we never went back to reconsider or test. We should.

I still need to write a test for the multi-DC scenario.
